Failure for MacOS - M3 Chip

Hello,

When I do `bundle install`, the gem called `absmartly-sdk` was throwing an error like below. `arraybuffer` is a dependency of the gem. The main issue was related to integer conversion and I was able to solve it with `bundle install -Wno-error=int-conversion` . 
I believe this is a special case for M3 chips and wanted to bring the issue to your attention.

If more clarification needed, please feel free to reach out.

Kind Regards,

```
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

...

compiling arraybuffer.c
arraybuffer.c:65:19: error: incompatible pointer to integer conversion assigning to 'VALUE' (aka 'unsigned long') from 'void *' [-Wint-conversion]
  bb->backing_str = NULL;
                  ^ ~~~~
arraybuffer.c:90:13: warning: pointer type mismatch ('char *' and 'char (*)[1]') [-Wpointer-type-mismatch]
  bb->ptr = BB_BACKING_PTR(bb);
            ^~~~~~~~~~~~~~~~~~
arraybuffer.c:74:50: note: expanded from macro 'BB_BACKING_PTR'
    (FL_TEST((bb)->backing_str, RSTRING_NOEMBED) ? RSTRING((bb)->backing_str)->as.heap.ptr : &RSTRING(bb->backing_str)->as.embed.ary)
                                                 ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
arraybuffer.c:115:3: warning: comparison of integers of different signs: 'int' and 'unsigned int' [-Wsign-compare]
  CHECKBOUNDS(bb, idx);
  ^~~~~~~~~~~~~~~~~~~~
arraybuffer.c:16:40: note: expanded from macro 'CHECKBOUNDS'
  if (!(bb)->ptr || (idx) < 0 || (idx) >= (bb)->size) { \
                                  ~~~  ^  ~~~~~~~~~~
arraybuffer.c:126:3: warning: comparison of integers of different signs: 'int' and 'unsigned int' [-Wsign-compare]
  CHECKBOUNDS(bb, idx);
  ^~~~~~~~~~~~~~~~~~~~
arraybuffer.c:16:40: note: expanded from macro 'CHECKBOUNDS'
  if (!(bb)->ptr || (idx) < 0 || (idx) >= (bb)->size) { \
                                  ~~~  ^  ~~~~~~~~~~
arraybuffer.c:142:23: warning: comparison of integers of different signs: 'int' and 'unsigned int' [-Wsign-compare]
    for (int i = 0; i < bb->size; i++) {
                    ~ ^ ~~~~~~~~
4 warnings and 1 error generated.
make: *** [arraybuffer.o] Error 1

make failed, exit code 2

...

An error occurred while installing arraybuffer (0.0.6), and Bundler cannot continue.

In Gemfile:
  absmartly-sdk was resolved to 1.0.7, which depends on
    arraybuffer
```

--- a/ext/arraybuffer/arraybuffer.c
+++ b/ext/arraybuffer/arraybuffer.c
@@ -13,7 +13,7 @@ extern VALUE cArrayBuffer;
   struct LLC_ArrayBuffer *bb = (struct LLC_ArrayBuffer*)rb_data_object_get((self))
 
 #define CHECKBOUNDS(bb, idx) \
-  if (!(bb)->ptr || (idx) < 0 || (idx) >= (bb)->size) { \
+  if (!(bb)->ptr || (idx) < 0 || (unsigned int)(idx) >= (bb)->size) { \
     rb_raise(rb_eArgError, "Index out of bounds: %d", (idx)); \
   }
 
@@ -62,7 +62,7 @@ t_bb_allocator(VALUE klass) {
   struct LLC_ArrayBuffer *bb = (struct LLC_ArrayBuffer*)xmalloc(sizeof(struct LLC_ArrayBuffer));
   bb->ptr = NULL;
   bb->size = 0;
-  bb->backing_str = NULL;
+  bb->backing_str = 0;
 
   return Data_Wrap_Struct(klass, t_bb_gc_mark, t_bb_free, bb);
 }
@@ -71,12 +71,12 @@ t_bb_allocator(VALUE klass) {
   // Ruby 3.1 and later
 
   #define BB_BACKING_PTR(bb) \
-    (FL_TEST((bb)->backing_str, RSTRING_NOEMBED) ? RSTRING((bb)->backing_str)->as.heap.ptr : &RSTRING(bb->backing_str)->as.embed.ary)
+    (FL_TEST((bb)->backing_str, RSTRING_NOEMBED) ? RSTRING((bb)->backing_str)->as.heap.ptr : RSTRING(bb->backing_str)->as.embed.ary)
 #else
   // Before ruby 3.1
 
   #define BB_BACKING_PTR(bb) \
-    (FL_TEST((bb)->backing_str, RSTRING_NOEMBED) ? RSTRING((bb)->backing_str)->as.heap.ptr : &RSTRING(bb->backing_str)->as.ary)
+    (FL_TEST((bb)->backing_str, RSTRING_NOEMBED) ? RSTRING((bb)->backing_str)->as.heap.ptr : RSTRING(bb->backing_str)->as.ary)
 #endif
 
 static void
@@ -87,7 +87,7 @@ t_bb_reassign_ptr(struct LLC_ArrayBuffer *bb) {
   }
 
   rb_str_set_len(bb->backing_str, bb->size);
-  bb->ptr = BB_BACKING_PTR(bb);
+  bb->ptr = (unsigned char*)BB_BACKING_PTR(bb);
   bb->ptr[bb->size] = 0;
 }
 
@@ -139,7 +139,7 @@ t_bb_each(VALUE self) {
   DECLAREBB(self);
 
   if (rb_block_given_p()) {
-    for (int i = 0; i < bb->size; i++) {
+    for (unsigned int i = 0; i < bb->size; i++) {
       unsigned int val = (unsigned int)bb->ptr[i];
       rb_yield(UINT2NUM(val));
     }

--- a/ext/arraybuffer/dataview.c
+++ b/ext/arraybuffer/dataview.c
@@ -131,8 +131,8 @@ t_dv_initialize(int argc, VALUE *argv, VALUE self) {
   if (offset_val < 0)
     rb_raise(rb_eArgError, "calculated offset is negative: %d", offset_val);
 
-  dv->offset = (unsigned long)offset_val;
-  dv->size = (unsigned long)size_val;
+  dv->offset = (unsigned int)offset_val;
+  dv->size = (unsigned int)size_val;
   dv->bb_obj = bb_obj;
 
   if (!keyword_ids[0]) {
@@ -214,7 +214,7 @@ t_dv_setoffset(VALUE self, VALUE offset) {
   if (offset_val < 0)
     rb_raise(rb_eArgError, "calculated offset is negative: %d", offset_val);
 
-  dv->offset = (unsigned long)offset_val;
+  dv->offset = (unsigned int)offset_val;
   return self;
 }
 
@@ -237,13 +237,13 @@ t_dv_setsize(VALUE self, VALUE size) {
   if (size_val < 0)
     rb_raise(rb_eArgError, "calculated size is negative: %d", size_val);
 
-  dv->size = (unsigned long)size_val;
+  dv->size = (unsigned int)size_val;
   return self;
 }
 
 #define DECLARENCHECKIDX(index) int idx = NUM2INT(index); \
   if (idx < 0) idx += (int)dv->size; \
-  if (idx < 0 || idx >= dv->size) rb_raise(rb_eArgError, "index out of bounds: %d", idx);
+  if (idx < 0 || idx >= (int)dv->size) rb_raise(rb_eArgError, "index out of bounds: %d", idx);
 
 #define CHECKBOUNDSBB(v) if ((v) < 0 || (v) >= (bb)->size) \
   rb_raise(rb_eArgError, "index out of underlying buffer bounds: %d", (v));
@@ -263,7 +263,7 @@ t_dv_getbit(VALUE self, VALUE index) {
   int idx = NUM2INT(index);
   if (idx < 0)
     idx += (int)dv->size * 8;
-  if (idx < 0 || idx >= dv->size * 8)
+  if (idx < 0 || idx >= (int)dv->size * 8)
     rb_raise(rb_eArgError, "index out of bounds: %d", idx);
 
   unsigned int bit_idx = ((unsigned int)idx) & 7;

